### ISSUE TITLE
Promote late variables to definitely assigned on a successful read.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1160]
Currently, a read of a potentially assigned late variable will either throw or succeed, but we don't use the knowledge that a success means that the variable was definitely assigned for anything.
We could promote the variable to definitely assigned on the non-throwing branch.

It's only for `late` variables that it makes a difference. A non-late variable cannot be attempted read while only potentially assigned. 
Reading a potentially assigned late variable is not a compile-time error, but it might throw at run-time instead, that's what `late` gives us. If the read doesn't throw, it's because the variable was assigned.

It would make `late final int x; ...potential assignment ...; x ??= 1;` an error because the assignment part of the `??=` happens after the read part, and then `x` is definitely assigned. It is a guaranteed run-time error, so making it a compile-time error is fine.

It would make `late int x; ...potential assignment...; x;` (or any read, including the one in `x??=1`) promote `x` to definitely assigned, as suggested here.

Promotion probably only works for local variables. Instance variables accessed through an interface are virtual, so we don't know that they are `late`, that's an implementation detail. Static (top-level and class) variables are resolved statically, so we know they are late, but we might not want to use that because it will make it a breaking change to switch a late field to a getter/setter pair with the same extrinsic behavior. So, local variables only, like all other promotions.

I don't think there is any corresponding feature for writing. Writing a non-final late variable is always OK, writing a final late variable either succeeds or throws, but in either case, it's definitely assigned afterwards. I guess we can promote to "definitely assigned" on the throw-branch, but I don't think we generally track throwing.

Promoting on read would be consistent with other promotions - we do promote `x` to non-nullable after an `x!` or `x as int`, checks that either succeed and prove the property, or throw.

(Based on https://dart-review.googlesource.com/c/sdk/+/157881/4/tests/language/nnbd/definite_assignment/definite_assignment_error_test.dart#453)

**Comment by eernstg** [comment]
Makes sense! I think this promotion is so similar to the one for `x!` and for `x as int` that it looks like a mistake that we don't have it.
